README.md
# vCard

[![pub package](https://img.shields.io/pub/v/vcard.svg)](https://pub.dartlang.org/packages/vcard)

Create vCards to import contacts into Outlook, iOS, Mac OS, and Android devices from your website or mobile application.

## Install

To use this package, add `vcard` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/) like so.

``` dart
dependencies:
  vcard: ^0.1.1
```

Then run the flutter tooling:
```sh
flutter packages get
```

## Usage

Below is a simple example of how to create a basic vCard and how to save it to a file, or view its contents from the console.

``` dart
/// Import the package
import 'package:vcard/vcard.dart';

///Create a new vCard
var vCard = VCard(
   firstName: 'FirstName',
   middleName: 'MiddleName',
   lastName: 'LastName',
   organization: 'ActivSpaces Labs',
   workPhone: 'Work Phone Number',
   birthday: DateTime.now(),
   title: 'Software Developer',
   url: 'https://github.com/valerycolong',
   note: 'Notes on contact'
  );

/// Save to file
vCard.saveToFile('./contact.vcf');

/// Get as formatted string
print(vCard.getFormattedString());

```

### Embedding Images

You can embed images in the photo or logo field instead of linking to them from a URL using base64 encoding.

```dart
//can be Windows or Linux/Unix path structures, and JPEG, PNG, GIF formats
vCard.photo.embedFromFile('/path/to/file.png');
vCard.logo.embedFromFile('/path/to/file.png');
```

```dart
//can also embed images via base-64 encoded strings
vCard.photo.embedFromString('iVBORw0KGgoAAAANSUhEUgAAA2...', 'image/png');
vCard.logo.embedFromString('iVBORw0KGgoAAAANSUhEUgAAA2...', 'image/png');
```

### Date Reference

MDN reference on how to use the `Date` object for birthday and anniversary can be found at [https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date).

### Complete Example

The following shows a vCard with everything filled out.

```dart
///import the package
import 'package:vcard/vcard.dart';

//create a new vCard
var vCard = VCard();

///set basic properties shown before
vCard.firstName = 'FirstName';
vCard.middleName = 'MiddleName';
vCard.lastName = 'Last Name';
vCard.uid = '6yuuhuhj-c34d-4a1e-8922-bd38a9476a53';
vCard.organization = 'ActivSpaces Labs';

///link to image
vCard.photo.attachFromUrl('/path/to/image/file.png', 'JPEG');

///or embed image
vCard.photo.attachFromUrl('/path/to/image/file.png');

vCard.workPhone = '312-555-1212';
vCard.birthday = DateTime.now();
vCard.title = 'Software Developer';
vCard.url = 'https://github.com/valerycolong';
vCard.workUrl = 'https://activspaces.com';
vCard.note = 'Notes on contact';

///set other vitals
vCard.nickname = 'Scarface';
vCard.namePrefix = 'Mr.';
vCard.nameSuffix = 'JR';
vCard.gender = 'M';
vCard.anniversary = DateTime.now();
vCard.role = 'Software Development';

///set other phone numbers
vCard.homePhone = 'Home Phone';
vCard.cellPhone = 'Cell Phone';
vCard.pagerPhone = 'Pager Phone';

///set fax/facsimile numbers
vCard.homeFax = 'Home Fax';
vCard.workFax = 'Work Fax';

///set email addresses
vCard.email = 'labs@activspaces.com';
vCard.workEmail = 'hello@activspaces.com';

///set logo of organization or personal logo (also supports embedding, see above)
vCard.logo.attachFromUrl('https://www.activspaces.com/wp-content/uploads/2019/01/ActivSpaces-Logo_Dark.png', 'PNG');

///set URL where the vCard can be found
vCard.source = 'http://example.com/myvcard.vcf';

///set address information
vCard.homeAddress.label = 'Home Address';
vCard.homeAddress.street = 'Great Soppo';
vCard.homeAddress.city = 'Buea';
vCard.homeAddress.stateProvince = 'SW';
vCard.homeAddress.postalCode = '00237';
vCard.homeAddress.countryRegion = 'Cameroon';
vCard.homeAddress.type = 'HOME';

vCard.workAddress.label = 'Work Address';
vCard.workAddress.street = 'Molyko';
vCard.workAddress.city = 'Buea';
vCard.workAddress.stateProvince = 'SW';
vCard.workAddress.postalCode = '00237';
vCard.workAddress.countryRegion = 'Cameroon';
vCard.workAddress.type = 'WORK';

///set social media URLs
vCard.socialUrls['facebook'] = 'https://...';
vCard.socialUrls['linkedIn'] = 'https://...';
vCard.socialUrls['twitter'] = 'https://...';
vCard.socialUrls['flickr'] = 'https://...';
vCard.socialUrls['custom'] = 'https://...';

///you can also embed photos from files instead of attaching via URL
vCard.photo.embedFromFile('photo.jpg');
vCard.logo.embedFromFile('logo.jpg');

vCard.version = '3.0'; //can also support 2.1 and 4.0, certain versions only support certain fields

///save to file
vCard.saveToFile('./contact/file.vcf');

///get as formatted string
print(vCard.getFormattedString());
```

### Multiple Email, Fax, & Phone Examples

`email`, `otherEmail`, `cellPhone`, `pagerPhone`, `homePhone`, `workPhone`, `homeFax`, `workFax`, `otherPhone` all support multiple entries in an array format.

Examples are provided below:

```dart

///import the package
import 'package:vcard/vcard.dart';


//create a new vCard
var vCard = VCard();

///multiple email entry
vCard.email = [
    'e.nesser@emailhost.tld',
    'e.nesser@emailhost2.tld',
    'e.nesser@emailhost3.tld'
];

///multiple cellphone
vCard.cellPhone = [
    '312-555-1414',
    '312-555-1415',
    '312-555-1416'
];

```

### Apple AddressBook Extensions

You can mark as a contact as an organization with the following Apple AddressBook extension property:

```dart
    var vCard = VCard();
    vCard.isOrganization = true;
```

## Original Source

This is a dart port of vCards-js package by Eric J Nesser
[https://github.com/enesser/vCards-js](https://github.com/enesser/vCards-js)

## Testing

You can run the VCard unit tests via (WIP):

```sh
flutter test
```

## Credits

Contributions are always welcome!

Thanks to Eric whose original work made life so much easier for me.
* Eric J Nesser ([https://github.com/enesser](https://github.com/enesser))

## Dependencies
This package depends on:
* validators ([https://pub.dev/packages/validators](https://pub.dev/packages/validators))
* path_provider ([https://pub.dev/packages/path_provider](https://pub.dev/packages/path_provider))
* path ([https://pub.dev/packages/path](https://pub.dev/packages/path))
* intl ([https://pub.dev/packages/intl](https://pub.dev/packages/intl))
* meta ([https://pub.dev/packages/meta](https://pub.dev/packages/meta))

## Contributions
Contributions are always welcome!

## Work with us
labs@activspaces.com

## License
Copyright (c)2019 ActivSpaces Labs MIT


CHANGELOG.md
## 0.1.2

* Upgraded dependency packages

## 0.1.1

* Upgraded dependency packages

## 0.1.0

* Format changes on vcard.dart

## 0.0.9

* Updated example README.md file.

## 0.0.8

* Updated example README.md file.

## 0.0.7

* Added example project to package.
* Remove use of console.log in examples in favour or print.

## 0.0.6

* Fixed minor formatting issues.

## 0.0.5

* Updated the package description.

## 0.0.4

* Minor changes with README.md file.

## 0.0.3

* Minor changes with README.md file.

## 0.0.2

* Added meta package dependency and small changes to README.md

## 0.0.1

* Initial Open Source release.


test/vcard_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:vcard/vcard.dart';

void main() {
  test('prints formated vcard string', () {
    final vcard = VCard();
    vcard.jobTitle = 'Dr';
    vcard.firstName = 'Colong';
    vcard.middleName = 'Valery';
    vcard.lastName = 'Nyiwung';
    vcard.homeAddress.street = 'Auberge';
    vcard.homeAddress.city = 'Buea';
    vcard.homeAddress.stateProvince = 'SW';
    vcard.homeAddress.countryRegion = 'Cameroon';
    //print(vcard.getFormattedString());
    expect(vcard.getFormattedString() is String, true);
    expect(vcard.getFormattedString().contains(RegExp(r'BEGIN:VCARD')), true);
  });
}


lib/vcard_formatter.dart
library vcard;

import 'package:intl/intl.dart';
import 'vcard.dart';

class VCardFormatter {
  int majorVersion = 3;

  /// Encode string
  /// @param  {String}     value to encode
  /// @return {String}     encoded string
  String e(String? value) {
    if ((value != null) && (value.isNotEmpty)) {
//      if (value is String) {
//        value = '' + value;
//      }
      return value
          .replaceAll(RegExp(r'/\n/g'), '\\n')
          .replaceAll(RegExp(r'/,/g'), '\\,')
          .replaceAll(RegExp(r'/;/g'), '\\;');
    }
    return '';
  }

  /// Return new line characters
  /// @return {String} new line characters
  String nl() {
    return '\r\n';
  }

  /// Get formatted photo
  /// @param  {String} photoType       Photo type (PHOTO, LOGO)
  /// @param  {String} url             URL to attach photo from
  /// @param  {String} mediaType       Media-type of photo (JPEG, PNG, GIF)
  /// @param  {String} isBase64        Whether or not is Base64 format
  /// @return {String}                 Formatted photo
  String getFormattedPhoto(String photoType, String url, String mediaType, bool isBase64) {
    String params;

    if (majorVersion >= 4) {
      params = isBase64 ? ';ENCODING=b;MEDIATYPE=image/' : ';MEDIATYPE=image/';
    } else if (majorVersion == 3) {
      params = isBase64 ? ';ENCODING=b;TYPE=' : ';TYPE=';
    } else {
      params = isBase64 ? ';ENCODING=BASE64;' : ';';
    }

    String formattedPhoto =
        photoType + params + mediaType + ':' + e(url) + nl();
    return formattedPhoto;
  }

  /// Get formatted address
  /// @param  {Map<String, String>}         address
  /// @param  {String}         type address type
  /// @param {String}         Encoding prefix encodingPrefix
  /// @return {String}         Formatted address
  String getFormattedAddress(
      {required MailingAddress address, required String encodingPrefix}) {
    var formattedAddress = '';

    if (address.label.isNotEmpty ||
        address.street.isNotEmpty ||
        address.city.isNotEmpty ||
        address.stateProvince.isNotEmpty ||
        address.postalCode.isNotEmpty ||
        address.countryRegion.isNotEmpty) {
      if (majorVersion >= 4) {
        formattedAddress = 'ADR' +
            encodingPrefix +
            ';TYPE=' +
            address.type +
            (address.label.isNotEmpty
                ? ';LABEL="' + e(address.label) + '"'
                : '') +
            ':;;' +
            e(address.street) +
            ';' +
            e(address.city) +
            ';' +
            e(address.stateProvince) +
            ';' +
            e(address.postalCode) +
            ';' +
            e(address.countryRegion) +
            nl();
      } else {
        if (address.label.isNotEmpty) {
          formattedAddress = 'LABEL' +
              encodingPrefix +
              ';TYPE=' +
              address.type +
              ':' +
              e(address.label) +
              nl();
        }
        formattedAddress += 'ADR' +
            encodingPrefix +
            ';TYPE=' +
            address.type +
            ':;;' +
            e(address.street) +
            ';' +
            e(address.city) +
            ';' +
            e(address.stateProvince) +
            ';' +
            e(address.postalCode) +
            ';' +
            e(address.countryRegion) +
            nl();
      }
    }

    return formattedAddress;
  }

  /// Convert date to YYYYMMDD format
  /// @param  {Date}       date to encode
  /// @return {String}     encoded date
  String formatVCardDate(DateTime date) {
    return DateFormat("yyyyMMdd").format(date);
  }

  String getFormattedString(VCard vCard) {
    majorVersion = vCard.getMajorVersion();

    String formattedVCardString = '';
    formattedVCardString += 'BEGIN:VCARD' + nl();
    formattedVCardString += 'VERSION:' + vCard.version + nl();

    String encodingPrefix = majorVersion >= 4 ? '' : ';CHARSET=UTF-8';
    String? formattedName = vCard.formattedName;

    if (formattedName == null) {
      formattedName = '';

      [vCard.firstName, vCard.middleName, vCard.lastName].forEach((name) {
        if ((name.isNotEmpty) && (formattedName!.isNotEmpty)) {
          formattedName = '$formattedName' + ' ';
        }
        //formattedName += name;
        formattedName = '$formattedName' + '$name';
      });
    }

    formattedVCardString +=
        'FN' + encodingPrefix + ':' + e(formattedName) + nl();
    formattedVCardString += 'N' +
        encodingPrefix +
        ':' +
        e(vCard.lastName) +
        ';' +
        e(vCard.firstName) +
        ';' +
        e(vCard.middleName) +
        ';' +
        e(vCard.namePrefix) +
        ';' +
        e(vCard.nameSuffix) +
        nl();

    if ((vCard.nickname.isNotEmpty) && (majorVersion >= 3)) {
      formattedVCardString +=
          'NICKNAME' + encodingPrefix + ':' + e(vCard.nickname) + nl();
    }

    if (vCard.gender != null) {
      formattedVCardString += 'GENDER:' + e(vCard.gender) + nl();
    }

    if (vCard.uid != null) {
      formattedVCardString +=
          'UID' + encodingPrefix + ':' + e(vCard.uid) + nl();
    }

    if (vCard.birthday != null) {
      formattedVCardString += 'BDAY:' + formatVCardDate(vCard.birthday!) + nl();
    }

    if ((vCard.anniversary != null) && (majorVersion >= 4)) {
      formattedVCardString +=
          'ANNIVERSARY:' + formatVCardDate(vCard.anniversary!) + nl();
    }

    if (vCard.email != null) {
      if (vCard.email is! List) {
        vCard.email = [vCard.email];
      }
      vCard.email.forEach((address) {
        if (majorVersion >= 4) {
          formattedVCardString +=
              'EMAIL' + encodingPrefix + ';type=HOME:' + e(address) + nl();
        } else if (majorVersion >= 3 && majorVersion < 4) {
          formattedVCardString += 'EMAIL' +
              encodingPrefix +
              ';type=HOME,INTERNET:' +
              e(address) +
              nl();
        } else {
          formattedVCardString +=
              'EMAIL' + encodingPrefix + ';HOME;INTERNET:' + e(address) + nl();
        }
      });
    }

    if (vCard.workEmail != null) {
      if (vCard.workEmail is! List) {
        vCard.workEmail = [vCard.workEmail];
      }
      vCard.workEmail.forEach((address) {
        if (majorVersion >= 4) {
          formattedVCardString +=
              'EMAIL' + encodingPrefix + ';type=WORK:' + e(address) + nl();
        } else if (majorVersion >= 3 && majorVersion < 4) {
          formattedVCardString += 'EMAIL' +
              encodingPrefix +
              ';type=WORK,INTERNET:' +
              e(address) +
              nl();
        } else {
          formattedVCardString +=
              'EMAIL' + encodingPrefix + ';WORK;INTERNET:' + e(address) + nl();
        }
      });
    }

    if (vCard.otherEmail != null) {
      if (vCard.otherEmail is! List) {
        vCard.otherEmail = [vCard.otherEmail];
      }
      vCard.otherEmail.forEach((address) {
        if (majorVersion >= 4) {
          formattedVCardString +=
              'EMAIL' + encodingPrefix + ';type=OTHER:' + e(address) + nl();
        } else if (majorVersion >= 3 && majorVersion < 4) {
          formattedVCardString += 'EMAIL' +
              encodingPrefix +
              ';type=OTHER,INTERNET:' +
              e(address) +
              nl();
        } else {
          formattedVCardString +=
              'EMAIL' + encodingPrefix + ';OTHER;INTERNET:' + e(address) + nl();
        }
      });
    }

    if (vCard.logo.url != null) {
      formattedVCardString += getFormattedPhoto(
          'LOGO', vCard.logo.url!, vCard.logo.mediaType!, vCard.logo.isBase64!);
    }

    if (vCard.photo.url != null) {
      formattedVCardString += getFormattedPhoto('PHOTO', vCard.photo.url!,
          vCard.photo.mediaType!, vCard.photo.isBase64!);
    }

    if (vCard.cellPhone != null) {
      if (vCard.cellPhone is! List) {
        vCard.cellPhone = [vCard.cellPhone];
      }

      vCard.cellPhone.forEach((number) {
        if (majorVersion >= 4) {
          formattedVCardString +=
              'TEL;VALUE=uri;TYPE="voice,cell":tel:' + e(number) + nl();
        } else {
          formattedVCardString += 'TEL;TYPE=CELL:' + e(number) + nl();
        }
      });
    }

    if (vCard.pagerPhone != null) {
      if (!vCard.pagerPhone is! List) {
        vCard.pagerPhone = [vCard.pagerPhone];
      }
      vCard.pagerPhone.forEach((number) {
        if (majorVersion >= 4) {
          formattedVCardString +=
              'TEL;VALUE=uri;TYPE="pager,cell":tel:' + e(number) + nl();
        } else {
          formattedVCardString += 'TEL;TYPE=PAGER:' + e(number) + nl();
        }
      });
    }

    if (vCard.homePhone != null) {
      if (vCard.homePhone is! List) {
        vCard.homePhone = [vCard.homePhone];
      }
      vCard.homePhone.forEach((number) {
        if (majorVersion >= 4) {
          formattedVCardString +=
              'TEL;VALUE=uri;TYPE="voice,home":tel:' + e(number) + nl();
        } else {
          formattedVCardString += 'TEL;TYPE=HOME,VOICE:' + e(number) + nl();
        }
      });
    }

    if (vCard.workPhone != null) {
      if (vCard.workPhone is! List) {
        vCard.workPhone = [vCard.workPhone];
      }
      vCard.workPhone.forEach((number) {
        if (majorVersion >= 4) {
          formattedVCardString +=
              'TEL;VALUE=uri;TYPE="voice,work":tel:' + e(number) + nl();
        } else {
          formattedVCardString += 'TEL;TYPE=WORK,VOICE:' + e(number) + nl();
        }
      });
    }

    if (vCard.homeFax != null) {
      if (vCard.homeFax is! List) {
        vCard.homeFax = [vCard.homeFax];
      }
      vCard.homeFax.forEach((number) {
        if (majorVersion >= 4) {
          formattedVCardString +=
              'TEL;VALUE=uri;TYPE="fax,home":tel:' + e(number) + nl();
        } else {
          formattedVCardString += 'TEL;TYPE=HOME,FAX:' + e(number) + nl();
        }
      });
    }

    if (vCard.workFax != null) {
      if (vCard.workFax is! List) {
        vCard.workFax = [vCard.workFax];
      }
      vCard.workFax.forEach((number) {
        if (majorVersion >= 4) {
          formattedVCardString +=
              'TEL;VALUE=uri;TYPE="fax,work":tel:' + e(number) + nl();
        } else {
          formattedVCardString += 'TEL;TYPE=WORK,FAX:' + e(number) + nl();
        }
      });
    }

    if (vCard.otherPhone != null) {
      if (vCard.otherPhone is! List) {
        vCard.otherPhone = [vCard.otherPhone];
      }
      vCard.otherPhone.forEach((number) {
        if (majorVersion >= 4) {
          formattedVCardString +=
              'TEL;VALUE=uri;TYPE="voice,other":tel:' + e(number) + nl();
        } else {
          formattedVCardString += 'TEL;TYPE=OTHER:' + e(number) + nl();
        }
      });
    }

    // Format Addresses
    formattedVCardString += getFormattedAddress(
        address: vCard.homeAddress, encodingPrefix: encodingPrefix);
    formattedVCardString += getFormattedAddress(
        address: vCard.workAddress, encodingPrefix: encodingPrefix);

    if (vCard.jobTitle != null) {
      formattedVCardString +=
          'TITLE' + encodingPrefix + ':' + e(vCard.jobTitle) + nl();
    }

    if (vCard.role != null) {
      formattedVCardString +=
          'ROLE' + encodingPrefix + ':' + e(vCard.role) + nl();
    }

    if (vCard.organization != null) {
      formattedVCardString +=
          'ORG' + encodingPrefix + ':' + e(vCard.organization) + nl();
    }

    if (vCard.url != null) {
      formattedVCardString +=
          'URL' + encodingPrefix + ':' + e(vCard.url) + nl();
    }

    if (vCard.workUrl != null) {
      formattedVCardString +=
          'URL;type=WORK' + encodingPrefix + ':' + e(vCard.workUrl) + nl();
    }

    if (vCard.note != null) {
      formattedVCardString +=
          'NOTE' + encodingPrefix + ':' + e(vCard.note) + nl();
    }

    vCard.socialUrls.forEach((key, value) {
      if ((value != null) && (value.isNotEmpty)) {
        formattedVCardString += 'X-SOCIALPROFILE' +
            encodingPrefix +
            ';TYPE=' +
            key +
            ':' +
            e(vCard.socialUrls[key]) +
            nl();
      }
    });

    if (vCard.source != null) {
      formattedVCardString +=
          'SOURCE' + encodingPrefix + ':' + e(vCard.source) + nl();
    }

    formattedVCardString += 'REV:' + DateTime.now().toIso8601String() + nl();

    if (vCard.isOrganization) {
      formattedVCardString += 'X-ABShowAs:COMPANY' + nl();
    }

    formattedVCardString += 'END:VCARD' + nl();
    return formattedVCardString;
  }
}


lib/vcard.dart
library vcard;

import 'dart:io';
import 'dart:convert';

import 'package:path/path.dart' as path;
import 'package:path_provider/path_provider.dart';
import 'package:validators/validators.dart';

import 'vcard_formatter.dart';

/// vCard, Valery Colong, May 2019

class VCard {
  /// Specifies a value that represents a persistent, globally unique identifier associated with the vCard
  String? uid;

  /// Date of birth
  DateTime? birthday;

  /// Anniversary
  DateTime? anniversary;

  /// Cell phone number
  var cellPhone;

  /// Other cell phone number or pager
  var pagerPhone;

  /// The address for private electronic mail communication
  var email;

  /// The address for work-related electronic mail communication
  var workEmail;

  /// Other email address
  var otherEmail;

  /// First name
  String firstName = '';

  /// Middle name
  String middleName = '';

  /// Last name
  String lastName = '';

  /// Formatted name string associated with the vCard object (will automatically populate if not set)
  String? formattedName;

  /// Prefix for individual's name
  String namePrefix = '';

  /// Suffix for individual's name
  String nameSuffix = '';

  /// Nickname of individual
  String nickname = '';

  /// Gender. Must be M or F for Male or Female
  String? gender;

  /// Home mailing address
  MailingAddress homeAddress = MailingAddress('HOME');

  /// Home phone
  var homePhone;

  /// Home facsimile
  var homeFax;

  /// Individual's photo
  Photo photo = Photo();

  /// Specifies supplemental information or a comment that is associated with the vCard
  String? note;

  /// The name and optionally the unit(s) of the organization associated with the vCard object
  String? organization;

  /// Logo
  Photo logo = Photo();

  /// The role, occupation, or business category of the vCard object within an organization
  String? role;

  /// Social URLs attached to the vCard object (ex: Facebook, Twitter, LinkedIn)
  /// @type {Map<String, String>}
  Map<String, String> socialUrls = getSocialUrls();

  /// A URL that can be used to get the latest version of this vCard
  String? source;

  /// Specifies the job title, functional position or function of the individual within an organization
  String? jobTitle;

  /// URL pointing to a website that represents the person in some way
  String? url;

  /// URL pointing to a website that represents the person's work in some way
  String? workUrl;

  /// Work mailing address
  MailingAddress workAddress = MailingAddress('WORK');

  /// Work phone
  var workPhone;

  /// Work facsimile
  var workFax;

  /// Other phone
  var otherPhone;

  /// Whether or not this vCard object represents an organization
  bool isOrganization = false;

  /// vCard version
  String version = '3.0';

  /// Get major version of the vCard format
  /// @return {integer}
  int getMajorVersion() {
    String majorVersionString =
        (this.version.isNotEmpty) ? this.version.split('.')[0] : '4';
    if (isNumeric(majorVersionString)) {
      return int.parse(majorVersionString);
    }
    return 4;
  }

  /// Get formatted vCard
  /// @return {String} Formatted vCard in VCF format
  String getFormattedString() => VCardFormatter().getFormattedString(this);

  /// Save formatted vCard to file
  /// @param  {String} filename
  saveToFile(filename) async {
    String contents = getFormattedString();

    final directory = await getApplicationDocumentsDirectory();
    final fs = File('${directory.path}/vCard/exports/$filename');
    fs.writeAsStringSync(contents);
  }

  /// Get social media URLs
  /// @return {Map<String, String>} Social media URL username map
  static Map<String, String> getSocialUrls() {
    return {
      'facebook': '',
      'linkedin': '',
      'twitter': '',
      'flickr': '',
      'instagram': '',
      'youtube': '',
    };
  }
}

class Photo {
  String? url;
  String? mediaType;
  bool? isBase64;

  /// Attach a photo from a URL
  /// @param  {string} url       URL where photo can be found
  /// @param  {string} mediaType Media type of photo (JPEG, PNG, GIF)
  attachFromUrl(url, mediaType) {
    this.url = url;
    this.mediaType = mediaType;
    this.isBase64 = false;
  }

  /// Embed a photo from a file using base-64 encoding
  /// @param  {string} filename
  embedFromFile(String imageFileUri) async {
    this.mediaType = path
        .extension(imageFileUri)
        .trim()
        .toUpperCase()
        .replaceFirst(RegExp(r'\./g'), '');
    final fs = File(imageFileUri);
    this.url = base64.encode(fs.readAsBytesSync());
    this.isBase64 = true;
  }

  /// Embed a photo from a base-64 string
  /// @param  {string} base64String
  embedFromString(base64String, mediaType) {
    this.mediaType = mediaType;
    this.url = base64String;
    this.isBase64 = true;
  }
}

class MailingAddress {
  /// Represents the actual text that should be put on the mailing label when delivering a physical package
  String label = '';

  /// Street address
  String street = '';

  /// City
  String city = '';

  /// State or province
  String stateProvince = '';

  /// Postal code
  String postalCode = '';

  /// Country or region
  String countryRegion = '';

  final String type;

  MailingAddress(this.type);
}
